Adds a warm start mechanism to the Run method to allow for online/batch training

bmcnns opened the pull request:
This update introduces a warm start mechanism to the GeneticProgrammingAlgorithm::Run method, enabling incremental or online updates. This enhancement aligns Operon with established frameworks such as PySR and Scikit-learn, significantly improving its flexibility and usability in iterative training workflows.

The addition of warm starting has immediate benefits for the PyOperon scikit-learn bindings, as it allows users to call .fit() incrementally.  This eliminates reliance on the callback interface and reduces associated limitations, making it much easier to use.

This feature is critical for my own research, where a single experiment may span up to 20 days without batching. With warm start and horizontal batching, the same experiment can be completed in a fraction of the time. The recent study ["A Survey on Batch Training in Genetic Programming" by Liah Rosenfeld and Leonardo Vanneschi](https://link.springer.com/article/10.1007/s10710-024-09501-6) highlights the promise of horizontal batching as a powerful strategy for scaling genetic programming.

[Moreover, warm start support is a requested feature in PyOperon](https://github.com/heal-research/pyoperon/issues/22), and a corresponding pull request has been opened in the PyOperon repository to integrate this functionality into the Python bindings.

foolnotion commented:
In addition to the comments above, there are a couple of more general issues which should be checked:

1. In a multi-threading context, the stop condition may be reached before every single thread gets the chance to finish generating and saving an offspring. We should make extra sure that the state of the algorithm after stop is consistent, and that neither `parents` or `offspring` spans contain uninitialized values (I think it should be fine, but worth checking).

2. We probably need some consistency checks for when a library user will call `Run(..., warmStart = true)` on an unstarted algorithm. Someone will do that just for fun and to watch it break.

3. The same changes should be applied to `nsga2.cpp`.

foolnotion commented on source/algorithms/gp.cpp:
```diff
@@ -65,11 +67,7 @@ auto GeneticProgrammingAlgorithm::Run(tf::Executor& executor, Operon::RandomGene
-                (*coeffInit)(rngs[i], parents[i].Genotype);
-            }).name("initialize population");
-            auto prepareEval = subflow.emplace([&]() { evaluator->Prepare(parents); }).name("prepare evaluator");
+	    auto prepareEval = subflow.emplace([&]() { evaluator->Prepare(parents); }).name("prepare evaluator");
```
inconsistent indentation, Operon uses 4 spaces and converts tabs to spaces.

bmcnns replied:
Fixed!

foolnotion commented on source/algorithms/gp.cpp:
```diff
@@ -79,9 +77,16 @@ auto GeneticProgrammingAlgorithm::Run(tf::Executor& executor, Operon::RandomGene
             prepareEval.precede(eval);
             eval.precede(reportProgress);
+
+	    if (!IsFitted() || !warmStart) {
```
Here as well, this line and the block should be aligned with the others

foolnotion commented on source/algorithms/nsga2.cpp:
```diff
@@ -156,10 +154,18 @@ auto NSGA2::Run(tf::Executor& executor, Operon::RandomGenerator& random, std::fu
             eval.precede(nonDominatedSort);
             nonDominatedSort.precede(reportProgress);
+
+	    if (!IsFitted() || !warmStart) {
```
inconsistent indentation (same as above)

bmcnns replied:
Fixed!

foolnotion commented on source/algorithms/nsga2.cpp:
```diff
@@ -156,10 +154,18 @@ auto NSGA2::Run(tf::Executor& executor, Operon::RandomGenerator& random, std::fu
             eval.precede(nonDominatedSort);
             nonDominatedSort.precede(reportProgress);
+
+	    if (!IsFitted() || !warmStart) {
```
Perhaps `if (!(isFitted() && warmStart))` would be slightly clearer.

bmcnns replied:
Agreed! changes made

bmcnns commented:
Hi @foolnotion , I think I've made all the requested changes:

- I've added a flag to ensure to prevent errors if a warm start is performed before a cold start
- I've added the equivalent logic to the NSGA2 algorithm
- I've additionally added a new method called RestoreIndividuals. I tried hard to avoid this but it was necessary due to constraints with pyoperon's Scikit-learn interface
- I've fixed the whitespace issues

I've updated the Pyoperon PR as well to add the workaround using RestoreIndividuals. Details are provided in that PR instead.

foolnotion commented on include/operon/algorithms/ga_base.hpp:
```diff
@@ -60,13 +60,23 @@ class GeneticAlgorithmBase {
     }
 
+    auto RestoreIndividuals(std::vector<Individual> inds) -> void
+    {
```
We should probably also check that the provided individuals vector is large enough:

```cpp
 EXPECT(inds.size() == config_.PoolSize + config_.PopulationSize,     
         "Mismatched number of individuals (must match pool/population sizes)");  
```

Of course, `>=` is also possible instead of `==` but this way we make sure that the user knows what they're doing.

foolnotion commented:
> Hi @foolnotion , I think I've made all the requested changes:
> 
>     * I've added a flag to ensure to prevent errors if a warm start is performed before a cold start
> 
>     * I've added the equivalent logic to the NSGA2 algorithm
> 
>     * I've additionally added a new method called RestoreIndividuals. I tried hard to avoid this but it was necessary due to constraints with pyoperon's Scikit-learn interface
> 
>     * I've fixed the whitespace issues
> 
> 
> I've updated the Pyoperon PR as well to add the workaround using RestoreIndividuals. Details are provided in that PR instead.

Hi,

Thanks for the changes and sorry for the long delay. The code looks good with one minor issue as highlighted above. Afterwards, we can merge this.

bmcnns commented:
Hi @foolnotion,

Likewise, sorry for the delay. I've made the change as requested. Happy to merge it if you are.

P.S. Looking forward to seeing you at GPTP!
